mithril-common/src/crypto_helper/cardano/key_certification.rs

```diff
@@ -0,0 +1,291 @@
+        let kes_signature =
+            Sum6KesSig::from_bytes(&bytes[256..]).map_err(|_| RegisterError::SerializationError)?;
+
+        Ok(Self {
+            stm_initializer,
+            kes_signature: Some(kes_signature),
+        })
+    }
+}
+
+impl KeyRegWrapper {
+    /// New Initialisation function. We temporarily keep the other init function,
+    /// but we should eventually transition to only use this one.
+    pub fn init(stake_dist: &ProtocolStakeDistribution) -> Self {
+        Self {
+            stm_key_reg: KeyReg::init(),
+            stake_distribution: HashMap::from_iter(stake_dist.to_vec()),
+        }
+    }
+
+    /// Register a new party. For a successful registration, the registrar needs to
+    /// provide the OpCert (in cbor form), the cold VK, a KES signature, and a
+    /// Mithril key (with its corresponding Proof of Possession).
+    pub fn register<P: AsRef<Path>>(
+        &mut self,
+        opcert_path: P,
+        kes_sig: ProtocolSignerVerificationKeySignature,
+        kes_period: usize,
+        pk: ProtocolSignerVerificationKey,
+    ) -> Result<(), RegisterError> {
+        let cert = OpCert::from_file(opcert_path)?;
+
+        cert.validate().map_err(|_| RegisterError::InvalidOpCert)?;
+
+        #[cfg(feature = "skip_signer_certification")]
+        println!("WARNING: Signer certification is skipped!!!");
+        #[cfg(feature = "skip_signer_certification")]
+        kes_sig
+            .verify(kes_period, &cert.kes_vk, &pk.to_bytes())
+            .map_err(|_| RegisterError::KesSignatureInvalid)?;
+
+        let mut hasher = Blake2b::<U28>::new();
+        hasher.update(cert.cold_vk.as_bytes());
+        let mut pool_id = [0u8; 28];
+        pool_id.copy_from_slice(hasher.finalize().as_slice());
+        let pool_id_bech32 = bech32::encode("pool", pool_id.to_base32(), Variant::Bech32)
+            .map_err(|_| RegisterError::InvalidOpCert)?; // TODO: maybe this should be a different error type like `RegisterError::InvalidPoolAddress`
+
+        if let Some(&stake) = self.stake_distribution.get(&pool_id_bech32) {
+            return self.stm_key_reg.register(stake, pk);
+        }
+
+        Err(RegisterError::KeyNonExisting)
+    }
+}
+
+impl StmSignerWrapper {
+    /// This function produces an STM signature
+    pub fn sign(&self, msg: &[u8]) -> Option<StmSig<D>> {
+        self.0.sign(msg)
+    }
+
+    /// This function should be called when a signing epoch is finished (or when a new one starts).
+    /// It consumes `self` and turns it back to an `StmInitializer`, which allows for an update in
+    /// the dynamic parameters (such as stake distribution, participants or KES signature). To ensure
+    /// that the `StmInitializer` will not be used for the previous registration, this function also
+    /// consumes the `ClosedKeyReg` instance. In case the stake of the current party has changed, it
+    /// includes it as input.
+    pub fn new_epoch(
```

Comment:
@iquerejeta this `new_epoch` function is not needed anymore as we rotate Signer secret keys at every epoch (you can also remove it from the `mithril-core` lib), as seen during our pairing session